[CC-17] Feat: Implement API for Job Posting Deletion

rumirifai opened the pull request:
Here's the pull request message reformatted with ## headings and horizontal separators.

## Description

This pull request addresses ticket **CC-17** and introduces the foundational API endpoint for soft deleting job postings.

This PR implements the `DELETE /api/v1/jobs/:uuid` endpoint, focusing on the core business logic for successful deletion scenarios, including handling for different job statuses (`draft`, `closed`) and special options (`force`, `preserveApplications`).

Note: This PR covers the "happy path" for deletions. Error handling, restoration, and other scenarios will be handled in subsequent PRs.

-----

## Changes Implemented

  - **[x] New Endpoint:** Created `DELETE /api/v1/jobs/:uuid` for soft deletion.
  - **[x] Authorization:** Secured the endpoint to ensure only authorized users (e.g., `Super Admin`, `Owner`) can perform deletions.
  - **[x] Service Layer Logic:** Implemented the core business logic in `JobDeletionService` for the following successful scenarios:
         - Deletion of draft jobs.
         - Deletion of closed jobs after the 7-day grace period.
         - Bypassing business rules when the `force=true` option is used by a Super Admin.
         - Handling the `preserveApplications=true` option.
  - **[x] Repository Layer:** Added the `softDeleteById` method to `JobPostingRepository`.
  - **[x] Testing:** Added unit tests for the successful deletion scenarios in `JobDeletionService` to ensure logic correctness.

-----

## How to Test Manually

1. Checkout this branch: git checkout `JOB-006-delete-api`
2. Install dependencies: `npm install`
3. Run database migrations: `npm run db:migrate`
4. Start the server: `npm run dev`
5. Using an API client (Postman/Thunder Client), obtain a JWT for a Super Admin user.
6. Create a `draft` job posting in the database and copy its uuid.
7. Send a DELETE request to `http://localhost:3000/api/v1/jobs/{uuid}` with the admin JWT.
8. Expected: A `200` response.
9. Verification: Check the `job_postings` table in the database. The `deleted_at` column for that job should now be filled with a timestamp.

-----

## Example Payloads

### Success Payload (200)

```json
{
  "status": 200,
  "message": "Job posting deleted successfully",
  "data": {
    "deleted_job": {
      "uuid": "string",
      "title": "string",
      "status": "deleted",
      "deleted_at": "datetime",
      "deleted_by": "integer"
    },
    "related_data": {
      "applications_preserved": 15,
      "notes_preserved": 8,
      "views_count_archived": 234
    },
    "recovery_info": {
      "recovery_possible": true,
      "recovery_deadline": "datetime (30 days from deletion)"
    }
  }
}
```


*Expected Result:* The request succeeds with a `200 OK` status. The response body will be a JSON object containing a success message and the updated data.

-----

## Related Ticket

**Closes:** [CC-17](https://revamp-radiosparx.atlassian.net/jira/software/projects/CC/boards/68?selectedIssue=CC-17)

rumirifai committed:
feat: tambahkan logika awal untuk hapus pekerjaan

rumirifai committed:
Feat: Add Job Deletion Logic and Routes

rumirifai committed:
Feat: Add unit testing for Successful Deletion

rumirifai committed:
Feat: Add department model from DEPT-001-department-api branch

rumirifai committed:
Feat: Add job category model from JOB-003-create-api branch

rumirifai committed:
Fix: Changes model and routes file

rumirifai committed:
Fix: Fix 27% duplication at job-deletion-service.test.ts

rumirifai committed:
Fix: Fix some files based on copilot review

rumirifai committed:
Fix: Fix some files based on SonarCube review

rumirifai committed:
Fix: Fix some files based on copilot review

rumirifai committed:
Fix: Fix testing problems

rumirifai committed:
Fix: Correct dependency in JobDeletionService

rumirifai committed:
Fix: Fix some files based on copilot review

rumirifai committed:
Fix: Fix some files based on SonarQube review

nugrahaAdhitama commented on src/repositories/job-posting-repository.ts:
```diff
@@ -0,0 +1,83 @@
+   * @param jobPostingUuid job posting uuid
+   * @returns true if there is at least one such application, false otherwise
+   */
+  async findWithActiveApplication(jobPostingUuid: string): Promise<boolean> {
+    const job = await this.findJobPostingByUuidWithSelect(jobPostingUuid, ['id', 'uuid']);
+    if (!job) return false;
+    
+    const application = await JobPostings.relatedQuery('jobApplications')
+      .for(job.id)
+      .whereIn('status', ['pending', 'under_review'])
+      .first();
+    return !!application;
+  }
```
The current implementation makes two separate database queries: one to get the job post and another to find related applications. This can be done in a single, more efficient query using whereExists.

**Reasoning**: You don't actually need any data from the JobPostings table itself, you just need to know if a record exists that matches both the UUID and the application status criteria. A subquery is perfect for this.

nugrahaAdhitama commented on src/repositories/job-posting-repository.ts:
```diff
@@ -0,0 +1,83 @@
+   * @param selectFields optional array of fields to select
+   * @returns JobPostingsData or null
+   */
+  private async findJobPostingByUuidWithSelect(jobPostingUuid: string, selectFields?: string[]) {
+    const query = JobPostings.query().findOne({ uuid: jobPostingUuid });
+    if (selectFields && selectFields.length > 0) {
+      query.select(...selectFields);
+    }
+    return await query;
+  }
```
The private findJobPostingByUuidWithSelect method adds a layer of abstraction that isn't necessary here. Since its only user (findWithActiveApplication) has been refactored to not need it, you can simplify findJobPostingByUuid directly.

Reasoning: Removing unused or overly-abstracted private methods makes the class smaller and easier to understand.

**SUGGESTION**
```
// Original
private async findJobPostingByUuidWithSelect(jobPostingUuid: string, selectFields?: string[]) {
  // ... implementation
}

async findJobPostingByUuid(jobPostingUuid: string): Promise<JobPostingsData | null> {
  const job = await this.findJobPostingByUuidWithSelect(jobPostingUuid);
  return job ?? null;
}

// Refactored
async findJobPostingByUuid(jobPostingUuid: string): Promise<JobPostingsData | null> {
  const job = await JobPostings.query().findOne({ uuid: jobPostingUuid });
  return job ?? null;
}
```

nugrahaAdhitama commented on src/repositories/job-posting-repository.ts:
```diff
@@ -0,0 +1,83 @@
+   * @param trx optional transaction
+   * @returns number of rows updated
+   */
+  async softDelete(jobPostingUuid: string, trx?: Transaction): Promise<number> {
+    const result = await JobPostings.query(trx)
+      .findOne({ uuid: jobPostingUuid })
+      .patch({ deleted_at: new Date() });
+    return result;
+  }
+
+  /**
+   * Restore a soft-deleted job posting (sets deleted_at to null)
+   * @param jobPostingUuid job posting uuid
+   * @param trx optional transaction
+   * @returns number of rows updated
+   */
+  async restore(jobPostingUuid: string, trx?: Transaction): Promise<number> {
+    const result = await JobPostings.query(trx)
+      .findOne({ uuid: jobPostingUuid })
+      .patch({ deleted_at: null });
+    return result;
+  }
+} 
```
The current findOne(...).patch(...) approach works, but a more direct way is to use patch(...).where(...). This guarantees a single UPDATE query without a preceding SELECT.

**Reasoning**: This is a more idiomatic and potentially more performant way to write update queries in Objection.js when you don't need the model instance beforehand. It translates directly to an UPDATE ... WHERE ... SQL statement.

**SUGGESTION**
```
// Original softDelete
async softDelete(jobPostingUuid: string, trx?: Transaction): Promise<number> {
  const result = await JobPostings.query(trx)
    .findOne({ uuid: jobPostingUuid })
    .patch({ deleted_at: new Date() });
  return result;
}

// Refactored softDelete
async softDelete(jobPostingUuid: string, trx?: Transaction): Promise<number> {
  return await JobPostings.query(trx)
    .where({ uuid: jobPostingUuid })
    .patch({ deleted_at: new Date() });
}

// Refactored restore
async restore(jobPostingUuid: string, trx?: Transaction): Promise<number> {
  return await JobPostings.query(trx)
    .where({ uuid: jobPostingUuid })
    .patch({ deleted_at: null });
}
```

nugrahaAdhitama commented on src/services/job-deletion-service.ts:
```diff
@@ -0,0 +1,216 @@
+     * @param option - DeleteJobOptions (force, preserveApplications)
+     * @returns DeleteJobResponse with detailed information about the deletion
+     */
+    async deleteJobPosting(jobPostingUuid: string, user: UserData, option: DeleteJobOptions = {}): Promise<DeleteJobResponse> {
+        const jobPosting = await this.repo.findJobPostingByUuid(jobPostingUuid);
+        if (!jobPosting || jobPosting.deleted_at) {
+            throw createNotFoundError('JobPosting', jobPostingUuid);
+        }
+
+        // Check authorization
+        this.validateDeletionAuthorization(jobPosting, user);
+
+        // Check business rules unless bypassed
+        const shouldBypassBusinessRules = option.force === true && this.isSuperAdmin(user);
+        if (!shouldBypassBusinessRules) {
+            await this.validateBusinessRules(jobPosting, jobPostingUuid, option);
+        }
+
+        // Perform soft delete
+        await this.repo.softDelete(jobPostingUuid);
+        
+        // Get updated job data and prepare response
+        return await this.prepareDeletionResponse(jobPostingUuid, user, jobPosting, option);
+    }
```
The deleteJobPosting method currently fetches the job data, and then prepareDeletionResponse fetches the exact same data a second time. We can avoid this second database call.

**CURRENT LOGIC**
```
// Perform soft delete
await this.repo.softDelete(jobPostingUuid);


// Get updated job data and prepare response
// This makes an unnecessary second DB call
return await this.prepareDeletionResponse(jobPostingUuid, user, jobPosting, option);
```

we can change it like this
```
// Perform soft delete
const deletionTimestamp = new Date();
await this.repo.softDelete(jobPostingUuid);

// Pass the original job data and the new timestamp to the response preparer
// This removes the need for a second database call
return this.prepareDeletionResponse(user, jobPosting, deletionTimestamp, option);
```

This also requires changing prepareDeletionResponse to be synchronous and accept the new parameters:
**BEFFORE**
```
private async prepareDeletionResponse(
    jobPostingUuid: string,
    user: UserData,
    originalJobPosting: any,
    option: DeleteJobOptions
): Promise<DeleteJobResponse> {
    const deletedJob = await this.repo.findJobPostingByUuid(jobPostingUuid);
// ...
```

**AFTER**
```
private prepareDeletionResponse(
    user: UserData,
    originalJobPosting: JobPostingsData,
    deletedAt: Date,
    option: DeleteJobOptions
): DeleteJobResponse {
    // No need to fetch the job again. We can build the response directly.
    return {
        deleted_job: {
            uuid: originalJobPosting.uuid,
            title: originalJobPosting.title,
            status: 'deleted',
            deleted_at: deletedAt, // Use the passed-in timestamp
            deleted_by: user.id,
        },
        // ...
    };
}
```

nugrahaAdhitama commented on src/services/job-deletion-service.ts:
```diff
@@ -0,0 +1,216 @@
+     * @param jobPostingUuid - Job posting UUID
+     * @param user - UserData performing the action
+     */
+    async restoreJobPosting(jobPostingUuid: string, user: UserData): Promise<void> {
+        const jobPosting = await this.repo.findJobPostingByUuid(jobPostingUuid);
+        if (!jobPosting) {
+            throw createNotFoundError('JobPosting', jobPostingUuid);
+        }
+        if (!jobPosting.deleted_at) {
+            throw createError(
+                ErrorType.RESOURCE_CONFLICT,
+                'Job posting is not deleted.'
+            );
+        }
+
+        // Authorization: Only Super Admin can restore
+        const isSuperAdmin = user.role?.name === 'Super Admin';
+        if (!isSuperAdmin) {
+            throw createError(
+                ErrorType.ACCESS_DENIED,
+                'Only Super Admin can restore job postings.'
+            );
+        }
+
+        // Business Rule: Only allow restore if deleted_at < 30 days ago
+        const deletedAt = new Date(jobPosting.deleted_at);
+        const now = new Date();
+        const diffDays = (now.getTime() - deletedAt.getTime()) / (1000 * 60 * 60 * 24);
+        if (diffDays > 30) {
+            throw createError(
+                ErrorType.RESOURCE_CONFLICT,
+                'Cannot restore job posting deleted more than 30 days ago.'
+            );
+        }
+
+        await this.repo.restore(jobPostingUuid);
+    }
```
The restoreJobPosting method contains a hardcoded "30" day rule and a direct check for the 'Super Admin' role. This logic is duplicated elsewhere or can be centralized for easier maintenance.

**CURRENT STATE**
```
// Authorization: Only Super Admin can restore
const isSuperAdmin = user.role?.name === 'Super Admin';
if (!isSuperAdmin) { /* ... */ }

// Business Rule: Only allow restore if deleted_at < 30 days ago
const deletedAt = new Date(jobPosting.deleted_at);
const now = new Date();
const diffDays = (now.getTime() - deletedAt.getTime()) / (1000 * 60 * 60 * 24);
if (diffDays > 30) { /* ... */ }
```

we can change it like this
```
// Add as a class property for consistency
private readonly RECOVERY_PERIOD_DAYS = 30;

// ... inside restoreJobPosting()
// Reuse the existing helper method for the authorization check
if (!this.isSuperAdmin(user)) {
    throw createError(
        ErrorType.ACCESS_DENIED,
        'Only Super Admin can restore job postings.'
    );
}

// Use the class constant for the business rule
const deletedAt = new Date(jobPosting.deleted_at);
const now = new Date();
const diffDays = (now.getTime() - deletedAt.getTime()) / (1000 * 60 * 60 * 24);
if (diffDays > this.RECOVERY_PERIOD_DAYS) {
    throw createError(
        ErrorType.RESOURCE_CONFLICT,
        `Cannot restore job posting deleted more than ${this.RECOVERY_PERIOD_DAYS} days ago.`
    );
}
```

nugrahaAdhitama reviewed (changes requested):
Never make `any`. Why make TypeScript if you end up giving everything `any`?

rumirifai committed:
refactor(repository): simplify and optimize job posting queries

- Simplify findJobPostingByUuid, softDelete, and restore methods to use direct queries and return results directly.
- Optimize findWithActiveApplication to use a single whereExists query with relatedQuery(jobApplications) for better efficiency.

rumirifai committed:
Remove unused import of Request

rumirifai committed:
Refactor: optimize job deletion queries and fix middleware type issues
